File: typescript/tests/test_provider.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";

import { X402Paywall } from "../src/provider.js";

const WALLET = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
const ROUTER = "0x887536bD817B758f99F090a80F48032a24f50916";
const USDC = "0x036CbD53842c5426634e7929541eC2318f3dCF7e";
const NETWORK = "eip155:31337";

function makePaywall(overrides: Partial<ConstructorParameters<typeof X402Paywall>[0]> = {}): X402Paywall {
  return new X402Paywall({
    walletAddress: WALLET,
    routerAddress: ROUTER,
    amountUsdc: 0.001,
    network: NETWORK,
    asset: USDC,
    facilitatorUrl: "http://localhost:3000",
    facilitatorToken: "test-token",
    ...overrides,
  });
}

function encodeSig(payload: unknown): string {
  return Buffer.from(JSON.stringify(payload)).toString("base64");
}

function makeDummyPaymentSig(): string {
  return encodeSig({
    scheme: "exact",
    network: NETWORK,
    x402Version: 1,
    payload: {
      signature: "0xdeadbeef",
      authorization: {
        from: "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
        to: WALLET,
        value: "1000",
        validAfter: "0",
        validBefore: "9999999999",
        nonce: "0xabcd1234",
      },
    },
  });
}

function mockFetch(response: { ok: boolean; json: () => unknown }): typeof globalThis.fetch {
  return async () =>
    ({
      ok: response.ok,
      json: async () => response.json(),
    }) as Response;
}

// ─── Construction ─────────────────────────────────────────────────────────────

describe("X402Paywall construction", () => {
  it("toJSON() reveals only non-sensitive fields", () => {
    const pw = makePaywall();
    const j = pw.toJSON();
    assert.equal(j.walletAddress, WALLET);
    assert.equal(j.network, NETWORK);
    // token must not appear
    assert.ok(!("facilitatorToken" in j));
  });
});

// ─── paymentRequiredHeader ────────────────────────────────────────────────────

describe("X402Paywall.paymentRequiredHeader", () => {
  it("produces correctly structured base64 JSON", () => {
    const pw = makePaywall({ amountUsdc: 0.005 });
    const raw = pw.paymentRequiredHeader();
    const payload = JSON.parse(Buffer.from(raw, "base64").toString("utf8")) as {
      scheme: string;
      network: string;
      amount: string;
      asset: string;
      payTo: string;
      maxTimeoutSeconds: number;
    };

    assert.equal(payload.scheme, "exact");
    assert.equal(payload.network, NETWORK);
    assert.equal(payload.amount, "5000"); // 0.005 USDC * 1_000_000
    assert.equal(payload.asset, USDC);
    assert.equal(payload.payTo, ROUTER);
    assert.ok(typeof payload.maxTimeoutSeconds === "number");
  });

  it("converts 0.001 USDC to 1000 base units", () => {
    const pw = makePaywall({ amountUsdc: 0.001 });
    const raw = pw.paymentRequiredHeader();
    const payload = JSON.parse(Buffer.from(raw, "base64").toString("utf8")) as { amount: string };
    assert.equal(payload.amount, "1000");
  });

  it("includes V2 resource/description/mimeType when configured", () => {
    const pw = makePaywall({
      resource: "/v1/data",
      description: "Market data feed",
      mimeType: "application/json",
    });
    const raw = pw.paymentRequiredHeader();
    const payload = JSON.parse(Buffer.from(raw, "base64").toString("utf8")) as Record<string, unknown>;
    assert.equal(payload["resource"], "/v1/data");
    assert.equal(payload["description"], "Market data feed");
    assert.equal(payload["mimeType"], "application/json");
  });

  it("omits V2 fields when not configured", () => {
    const pw = makePaywall();
    const raw = pw.paymentRequiredHeader();
    const payload = JSON.parse(Buffer.from(raw, "base64").toString("utf8")) as Record<string, unknown>;
    assert.ok(!("resource" in payload), "resource must be absent");
    assert.ok(!("description" in payload), "description must be absent");
    assert.ok(!("mimeType" in payload), "mimeType must be absent");
  });
});

// ─── check ─────────────────────────────────────────────────────────────────────

describe("X402Paywall.check", () => {
  it("returns { isValid: false } when payment_sig is null", async () => {
    const pw = makePaywall();
    const result = await pw.check(null);
    assert.equal(result.isValid, false);
    assert.equal(result.invalidReason, undefined);
  });

  it("returns INVALID_PAYLOAD for malformed base64", async () => {
    const pw = makePaywall();
    const result = await pw.check("!!!not-valid-base64!!!");
    assert.equal(result.isValid, false);
    assert.equal(result.invalidReason, "INVALID_PAYLOAD");
  });

  it("returns { isValid: true } when facilitator says valid", async () => {
    const pw = makePaywall();
    const orig = globalThis.fetch;
    try {
      (globalThis as Record<string, unknown>)["fetch"] = mockFetch({ ok: true, json: () => ({ isValid: true }) });
      const result = await pw.check(makeDummyPaymentSig());
      assert.equal(result.isValid, true);
    } finally {
      globalThis.fetch = orig;
    }
  });

  it("returns invalidReason when facilitator says invalid", async () => {
    const pw = makePaywall();
    const orig = globalThis.fetch;
    try {
      (globalThis as Record<string, unknown>)["fetch"] = mockFetch({
        ok: true,
        json: () => ({ isValid: false, invalidReason: "SIGNATURE_INVALID" }),
      });
      const result = await pw.check(makeDummyPaymentSig());
      assert.equal(result.isValid, false);
      assert.equal(result.invalidReason, "SIGNATURE_INVALID");
    } finally {
      globalThis.fetch = orig;
    }
  });

  it("returns FACILITATOR_ERROR when fetch throws", async () => {
    const pw = makePaywall();
    const orig = globalThis.fetch;
    try {
      (globalThis as Record<string, unknown>)["fetch"] = async () => { throw new Error("network error"); };
      const result = await pw.check(makeDummyPaymentSig());
      assert.equal(result.isValid, false);
      assert.equal(result.invalidReason, "FACILITATOR_ERROR");
    } finally {
      globalThis.fetch = orig;
    }
  });

  it("returns FACILITATOR_ERROR when facilitator returns non-ok status", async () => {
    const pw = makePaywall();
    const orig = globalThis.fetch;
    try {
      (globalThis as Record<string, unknown>)["fetch"] = mockFetch({ ok: false, json: () => ({}) });
      const result = await pw.check(makeDummyPaymentSig());
      assert.equal(result.isValid, false);
      assert.equal(result.invalidReason, "FACILITATOR_ERROR");
    } finally {
      globalThis.fetch = orig;
    }
  });

  it("sends Authorization header when token is configured", async () => {
    const pw = makePaywall({ facilitatorToken: "my-jwt" });
    let capturedHeaders: Record<string, string> = {};
    const orig = globalThis.fetch;
    try {
      (globalThis as Record<string, unknown>)["fetch"] = async (_url: unknown, init?: RequestInit) => {
        capturedHeaders = Object.fromEntries(new Headers(init?.headers).entries());
        return { ok: true, json: async () => ({ isValid: true }) } as Response;
      };
      await pw.check(makeDummyPaymentSig());
    } finally {
      globalThis.fetch = orig;
    }
    assert.equal(capturedHeaders["authorization"], "Bearer my-jwt");
  });

  it("omits Authorization header when no token", async () => {
    const pw = new X402Paywall({ walletAddress: WALLET, routerAddress: ROUTER, amountUsdc: 0.001, network: NETWORK, asset: USDC });
    let capturedHeaders: Record<string, string> = {};
    const orig = globalThis.fetch;
    try {
      (globalThis as Record<string, unknown>)["fetch"] = async (_url: unknown, init?: RequestInit) => {
        capturedHeaders = Object.fromEntries(new Headers(init?.headers).entries());
        return { ok: true, json: async () => ({ isValid: true }) } as Response;
      };
      await pw.check(makeDummyPaymentSig());
    } finally {
      globalThis.fetch = orig;
    }
    assert.ok(!("authorization" in capturedHeaders));
  });
});

// ─── handle ────────────────────────────────────────────────────────────────────

describe("X402Paywall.handle", () => {
  it("returns 402 Response when PAYMENT-SIGNATURE header is absent", async () => {
    const pw = makePaywall();
    const req = new Request("http://example.com/v1/data");
    const orig = globalThis.fetch;
    try {
      // check() will return false (null sig) without hitting facilitator
      const resp = await pw.handle(req);
      assert.ok(resp instanceof Response);
      assert.equal(resp!.status, 402);
      assert.ok(resp!.headers.get("payment-required"));
    } finally {
      globalThis.fetch = orig;
    }
  });

  it("returns null when payment is valid", async () => {
    const pw = makePaywall();
    const req = new Request("http://example.com/v1/data", {
      headers: { "payment-signature": makeDummyPaymentSig() },
    });
    const orig = globalThis.fetch;
    try {
      (globalThis as Record<string, unknown>)["fetch"] = mockFetch({ ok: true, json: () => ({ isValid: true }) });
      const resp = await pw.handle(req);
      assert.equal(resp, null);
    } finally {
      globalThis.fetch = orig;
    }
  });
});

// ─── honoMiddleware ────────────────────────────────────────────────────────────

describe("X402Paywall.honoMiddleware", () => {
  function makeHonoContext(paymentSig?: string): {
    req: { raw: Request };
    header(name: string, value: string): void;
    body(content: string, status?: number): Response;
  } {
    const headers = new Headers();
    if (paymentSig) headers.set("payment-signature", paymentSig);
    const raw = new Request("http://example.com/v1/data", { headers });
    const resHeaders: Record<string, string> = {};
    return {
      req: { raw },
      header(name: string, value: string) { resHeaders[name] = value; },
      body(content: string, status = 200): Response {
        return new Response(content, { status, headers: resHeaders });
      },
    };
  }

  it("returns 402 Response when payment is absent", async () => {
    const pw = makePaywall();
    const middleware = pw.honoMiddleware();
    const c = makeHonoContext();
    let nextCalled = false;
    const resp = await middleware(c, async () => { nextCalled = true; });
    assert.ok(resp instanceof Response, "should return a Response");
    assert.equal((resp as Response).status, 402);
    assert.ok(!nextCalled, "next should not be called");
  });

  it("calls next when payment is valid", async () => {
    const pw = makePaywall();
    const middleware = pw.honoMiddleware();
    const c = makeHonoContext(makeDummyPaymentSig());
    let nextCalled = false;
    const orig = globalThis.fetch;
    try {
      (globalThis as Record<string, unknown>)["fetch"] = mockFetch({ ok: true, json: () => ({ isValid: true }) });
      const resp = await middleware(c, async () => { nextCalled = true; });
      assert.equal(resp, undefined);
      assert.ok(nextCalled, "next must be called when payment is valid");
    } finally {
      globalThis.fetch = orig;
    }
  });
});

File: python/tests/test_provider.py
"""Tests for X402Paywall service provider middleware."""

from __future__ import annotations

import base64
import json
from unittest.mock import AsyncMock, MagicMock, patch

import pytest

from remitmd.provider import X402Paywall

_WALLET = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"
_USDC = "0x142aD61B8d2edD6b3807D9266866D97C35Ee0317"
_NETWORK = "eip155:31337"


def _make_paywall(**kwargs: object) -> X402Paywall:
    return X402Paywall(
        wallet_address=_WALLET,
        amount_usdc=kwargs.pop("amount_usdc", 0.001),  # type: ignore[arg-type]
        network=kwargs.pop("network", _NETWORK),  # type: ignore[arg-type]
        asset=kwargs.pop("asset", _USDC),  # type: ignore[arg-type]
        facilitator_url=kwargs.pop("facilitator_url", "http://localhost:3000"),  # type: ignore[arg-type]
        facilitator_token=kwargs.pop("facilitator_token", "test-token"),  # type: ignore[arg-type]
        **kwargs,
    )


def _encode_payload(payload: dict[str, object]) -> str:
    return base64.b64encode(json.dumps(payload).encode()).decode()


# ─── Construction ─────────────────────────────────────────────────────────────


def test_paywall_repr() -> None:
    pw = _make_paywall()
    assert _WALLET in repr(pw)
    assert _NETWORK in repr(pw)


def test_amount_converted_to_base_units() -> None:
    pw = X402Paywall(
        wallet_address=_WALLET,
        amount_usdc=0.001,
        network=_NETWORK,
        asset=_USDC,
    )
    header = json.loads(base64.b64decode(pw.payment_required_header()))
    assert header["amount"] == "1000"  # 0.001 USDC * 1_000_000


# ─── payment_required_header ──────────────────────────────────────────────────


def test_payment_required_header_structure() -> None:
    pw = _make_paywall(amount_usdc=0.005)
    raw = pw.payment_required_header()
    payload = json.loads(base64.b64decode(raw))

    assert payload["scheme"] == "exact"
    assert payload["network"] == _NETWORK
    assert payload["amount"] == "5000"  # 0.005 USDC
    assert payload["asset"] == _USDC
    assert payload["payTo"] == _WALLET
    assert isinstance(payload["maxTimeoutSeconds"], int)


def test_payment_required_header_v2_fields() -> None:
    from remitmd.provider import X402Paywall

    pw = X402Paywall(
        wallet_address=_WALLET,
        amount_usdc=0.001,
        network=_NETWORK,
        asset=_USDC,
        resource="/v1/data",
        description="Market data feed",
        mime_type="application/json",
    )
    payload = json.loads(base64.b64decode(pw.payment_required_header()))
    assert payload["resource"] == "/v1/data"
    assert payload["description"] == "Market data feed"
    assert payload["mimeType"] == "application/json"


def test_payment_required_header_v2_fields_absent_by_default() -> None:
    pw = _make_paywall()
    payload = json.loads(base64.b64decode(pw.payment_required_header()))
    assert "resource" not in payload
    assert "description" not in payload
    assert "mimeType" not in payload


# ─── check — no signature ─────────────────────────────────────────────────────


@pytest.mark.asyncio
async def test_check_returns_false_when_no_signature() -> None:
    pw = _make_paywall()
    is_valid, reason = await pw.check(None)
    assert is_valid is False
    assert reason is None  # absent header, not an error


@pytest.mark.asyncio
async def test_check_returns_false_for_empty_string() -> None:
    pw = _make_paywall()
    is_valid, reason = await pw.check("")
    assert is_valid is False
    assert reason is None


# ─── check — malformed payload ────────────────────────────────────────────────


@pytest.mark.asyncio
async def test_check_returns_false_for_non_base64() -> None:
    pw = _make_paywall()
    is_valid, reason = await pw.check("not-valid-base64!!!")
    assert is_valid is False
    assert reason == "INVALID_PAYLOAD"


# ─── check — facilitator call ─────────────────────────────────────────────────


def _make_dummy_payment_sig() -> str:
    """Build a syntactically valid (but not cryptographically verified) PAYMENT-SIGNATURE."""
    payload = {
        "scheme": "exact",
        "network": _NETWORK,
        "x402Version": 1,
        "payload": {
            "signature": "0xdeadbeef",
            "authorization": {
                "from": "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
                "to": _WALLET,
                "value": "1000",
                "validAfter": "0",
                "validBefore": "9999999999",
                "nonce": "0xabcd1234",
            },
        },
    }
    return base64.b64encode(json.dumps(payload).encode()).decode()


@pytest.mark.asyncio
async def test_check_calls_facilitator_and_returns_valid() -> None:
    pw = _make_paywall()
    payment_sig = _make_dummy_payment_sig()

    mock_response = MagicMock()
    mock_response.raise_for_status = MagicMock()
    mock_response.json = MagicMock(return_value={"isValid": True})

    with patch("httpx.AsyncClient") as mock_client_class:
        mock_client = AsyncMock()
        mock_client_class.return_value.__aenter__ = AsyncMock(return_value=mock_client)
        mock_client_class.return_value.__aexit__ = AsyncMock(return_value=None)
        mock_client.post = AsyncMock(return_value=mock_response)

        is_valid, reason = await pw.check(payment_sig)

    assert is_valid is True
    assert reason is None

    # Verify the facilitator was called with correct body.
    call_kwargs = mock_client.post.call_args
    assert "/api/v0/x402/verify" in call_kwargs.args[0]
    body = call_kwargs.kwargs["json"]
    assert body["paymentRequired"]["payTo"] == _WALLET
    assert body["paymentRequired"]["amount"] == "1000"
    assert body["paymentRequired"]["network"] == _NETWORK
    assert call_kwargs.kwargs["headers"].get("Authorization") == "Bearer test-token"


@pytest.mark.asyncio
async def test_check_returns_false_when_facilitator_says_invalid() -> None:
    pw = _make_paywall()
    payment_sig = _make_dummy_payment_sig()

    mock_response = MagicMock()
    mock_response.raise_for_status = MagicMock()
    mock_response.json = MagicMock(
        return_value={"isValid": False, "invalidReason": "SIGNATURE_INVALID"}
    )

    with patch("httpx.AsyncClient") as mock_client_class:
        mock_client = AsyncMock()
        mock_client_class.return_value.__aenter__ = AsyncMock(return_value=mock_client)
        mock_client_class.return_value.__aexit__ = AsyncMock(return_value=None)
        mock_client.post = AsyncMock(return_value=mock_response)

        is_valid, reason = await pw.check(payment_sig)

    assert is_valid is False
    assert reason == "SIGNATURE_INVALID"


@pytest.mark.asyncio
async def test_check_returns_facilitator_error_on_exception() -> None:
    pw = _make_paywall()
    payment_sig = _make_dummy_payment_sig()

    with patch("httpx.AsyncClient") as mock_client_class:
        mock_client = AsyncMock()
        mock_client_class.return_value.__aenter__ = AsyncMock(return_value=mock_client)
        mock_client_class.return_value.__aexit__ = AsyncMock(return_value=None)
        mock_client.post = AsyncMock(side_effect=Exception("connection refused"))

        is_valid, reason = await pw.check(payment_sig)

    assert is_valid is False
    assert reason == "FACILITATOR_ERROR"


@pytest.mark.asyncio
async def test_check_omits_auth_header_when_no_token() -> None:
    pw = X402Paywall(
        wallet_address=_WALLET,
        amount_usdc=0.001,
        network=_NETWORK,
        asset=_USDC,
        facilitator_url="http://localhost:3000",
    )
    payment_sig = _make_dummy_payment_sig()

    mock_response = MagicMock()
    mock_response.raise_for_status = MagicMock()
    mock_response.json = MagicMock(return_value={"isValid": True})

    with patch("httpx.AsyncClient") as mock_client_class:
        mock_client = AsyncMock()
        mock_client_class.return_value.__aenter__ = AsyncMock(return_value=mock_client)
        mock_client_class.return_value.__aexit__ = AsyncMock(return_value=None)
        mock_client.post = AsyncMock(return_value=mock_response)

        await pw.check(payment_sig)

    headers = mock_client.post.call_args.kwargs["headers"]
    assert "Authorization" not in headers
